### PR TITLE
[Core] Support TableScan resources clean

### LINKED ISSUE #2122: [Feature] Support TableScan.close to clean scan resources
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

We'll measure scan operations by introducing scan metrics, the metric groups will be in the JVM. In the resident JVM like flink session cluster, the metric groups should be cleaned in the right time. Like scan resources, we should clean them by SplitEnumerator closes.

### Solution

Add close() method in the TableScan interface first.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by schnappi17**
### Purpose

Linked issue: close #2122

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/table/source/TableScan.java [hunk 1]
```diff
@@ -47,4 +47,7 @@ public interface TableScan {
     interface Plan {
         List<Split> splits();
     }
+
+    /** Close this scan, clean resources here. */
+    void close();
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/table/source/TableScan.java:52]: `TableScan` is a `@Public` interface, do we really need to add a `close()` method here?

File: paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java [hunk 2]
```diff
@@ -187,6 +187,11 @@ protected StartingScanner createStartingScanner(boolean isStreaming) {
         }
     }
 
+    @Override
+    public void close() {
+        // do nothing yet
```
[the remark this answers is not in the document]
Comment (schnappi17) [reply paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java:192]: > All the `close()` methods in `AbstractInnerTableScan` and `ReadOnceTableScan` are empty, why do we need it?

@FangYongs This is a base pr, the scan metrics will rebase this pr and do substantial metrics resource clean in the close method.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Hi @schnappi17 

I think we should avoid adding `close` for metrics only.
1. This may lead to lots of closes in our codes, looks very tricky.
2. This breaks back Backwards Compatibility.

We may need to revisit `Metrics` design. I know Hudi also use static `Metrics` field to maintain all metrics, but when I take a look to its implementation. It is very tricky. It never delete metrics, and the shutdown of metrics is very casualness.

When I take a look to the metrics of Flink: https://issues.apache.org/jira/browse/FLINK-7876  It looks very good, Paimon, may not need to manage metrics ourselves, just register metrics externally through an interface.
For example, registering with the metrics system of a computing engine, or injecting users into their own metrics system.

The interface can just be a class like `MetricRegistry`.

**Comment by schnappi17** [comment]
Thanks @JingsongLi and @FangYongs for reviewing and suggestions, I'll take a look at this and thinking about the metrics interfaces design.

**Pull request closed without merging**